File: internal/utils/time.go
// Package utils provides utility functions for the Ithil application.
package utils

import (
	"fmt"
	"time"
)

// FormatTimestamp formats a timestamp for display.
// If relative is true, it returns a relative time string (e.g., "5m ago").
// Otherwise, it returns an absolute time string.
func FormatTimestamp(t time.Time, relative bool) string {
	if relative {
		return FormatRelativeTime(t)
	}
	return FormatAbsoluteTime(t)
}

// FormatRelativeTime formats a time as a relative string (e.g., "5m ago", "2h ago").
func FormatRelativeTime(t time.Time) string {
	now := time.Now()
	diff := now.Sub(t)

	// Future times
	if diff < 0 {
		diff = -diff
		switch {
		case diff < time.Minute:
			return "in a moment"
		case diff < time.Hour:
			minutes := int(diff.Minutes())
			return fmt.Sprintf("in %dm", minutes)
		case diff < 24*time.Hour:
			hours := int(diff.Hours())
			return fmt.Sprintf("in %dh", hours)
		default:
			return FormatAbsoluteTime(t)
		}
	}

	// Past times
	switch {
	case diff < time.Minute:
		return "just now"
	case diff < time.Hour:
		minutes := int(diff.Minutes())
		if minutes == 1 {
			return "1m ago"
		}
		return fmt.Sprintf("%dm ago", minutes)
	case diff < 24*time.Hour:
		hours := int(diff.Hours())
		if hours == 1 {
			return "1h ago"
		}
		return fmt.Sprintf("%dh ago", hours)
	case diff < 7*24*time.Hour:
		days := int(diff.Hours() / 24)
		if days == 1 {
			return "yesterday"
		}
		return fmt.Sprintf("%dd ago", days)
	case diff < 30*24*time.Hour:
		weeks := int(diff.Hours() / 24 / 7)
		if weeks == 1 {
			return "1w ago"
		}
		return fmt.Sprintf("%dw ago", weeks)
	default:
		return FormatAbsoluteTime(t)
	}
}

// FormatAbsoluteTime formats a time as an absolute string.
func FormatAbsoluteTime(t time.Time) string {
	now := time.Now()

	// If today, show only time
	if isToday(t, now) {
		return t.Format("15:04")
	}

	// If yesterday
	if isYesterday(t, now) {
		return "Yesterday " + t.Format("15:04")
	}

	// If this year, show date without year
	if t.Year() == now.Year() {
		return t.Format("Jan 02 15:04")
	}

	// Otherwise, show full date
	return t.Format("Jan 02, 2006 15:04")
}

// FormatTime12Hour formats a time in 12-hour format.
func FormatTime12Hour(t time.Time) string {
	return t.Format("3:04 PM")
}

// FormatTime24Hour formats a time in 24-hour format.
func FormatTime24Hour(t time.Time) string {
	return t.Format("15:04")
}

// FormatDate formats a date.
func FormatDate(t time.Time) string {
	now := time.Now()

	if isToday(t, now) {
		return "Today"
	}

	if isYesterday(t, now) {
		return "Yesterday"
	}

	if t.Year() == now.Year() {
		return t.Format("January 2")
	}

	return t.Format("January 2, 2006")
}

// FormatDateTime formats a date and time.
func FormatDateTime(t time.Time) string {
	now := time.Now()

	dateStr := FormatDate(t)
	timeStr := t.Format("15:04")

	if isToday(t, now) {
		return timeStr
	}

	return dateStr + " at " + timeStr
}

// FormatDuration formats a duration in a human-readable way.
func FormatDuration(d time.Duration) string {
	if d < time.Minute {
		seconds := int(d.Seconds())
		return fmt.Sprintf("%ds", seconds)
	}

	if d < time.Hour {
		minutes := int(d.Minutes())
		seconds := int(d.Seconds()) % 60
		if seconds == 0 {
			return fmt.Sprintf("%dm", minutes)
		}
		return fmt.Sprintf("%dm %ds", minutes, seconds)
	}

	hours := int(d.Hours())
	minutes := int(d.Minutes()) % 60
	if minutes == 0 {
		return fmt.Sprintf("%dh", hours)
	}
	return fmt.Sprintf("%dh %dm", hours, minutes)
}

// FormatLastSeen formats the last seen time for a user.
func FormatLastSeen(t time.Time) string {
	now := time.Now()
	diff := now.Sub(t)

	if diff < time.Minute {
		return "online"
	}

	if diff < 5*time.Minute {
		return "last seen recently"
	}

	if diff < time.Hour {
		minutes := int(diff.Minutes())
		return fmt.Sprintf("last seen %d minutes ago", minutes)
	}

	if diff < 24*time.Hour {
		hours := int(diff.Hours())
		if hours == 1 {
			return "last seen 1 hour ago"
		}
		return fmt.Sprintf("last seen %d hours ago", hours)
	}

	return "last seen " + FormatDate(t)
}

// isToday checks if a time is today.
func isToday(t, now time.Time) bool {
	tYear, tMonth, tDay := t.Date()
	nowYear, nowMonth, nowDay := now.Date()
	return tYear == nowYear && tMonth == nowMonth && tDay == nowDay
}

// isYesterday checks if a time is yesterday.
func isYesterday(t, now time.Time) bool {
	yesterday := now.AddDate(0, 0, -1)
	tYear, tMonth, tDay := t.Date()
	yYear, yMonth, yDay := yesterday.Date()
	return tYear == yYear && tMonth == yMonth && tDay == yDay
}

// ParseDuration parses a duration string like "5m", "2h", "1d".
func ParseDuration(s string) (time.Duration, error) {
	return time.ParseDuration(s)
}

File: src/utils/time.rs
//! Time formatting utilities.
//!
//! This module provides functions for formatting timestamps and durations
//! in human-readable formats.

use chrono::{DateTime, Duration, Local, Utc};

/// Formats a timestamp for display.
///
/// # Arguments
///
/// * `time` - The timestamp to format
/// * `relative` - If true, returns relative time (e.g., "5m ago"); otherwise absolute
///
/// # Examples
///
/// ```
/// use chrono::Utc;
/// use ithil::utils::format_timestamp;
///
/// let now = Utc::now();
/// let result = format_timestamp(now, true);
/// assert_eq!(result, "just now");
/// ```
#[must_use]
pub fn format_timestamp(time: DateTime<Utc>, relative: bool) -> String {
    if relative {
        format_relative_time(time)
    } else {
        format_absolute_time(time)
    }
}

/// Formats a time as a relative string (e.g., "5m ago", "2h ago").
///
/// # Arguments
///
/// * `time` - The timestamp to format
///
/// # Returns
///
/// A human-readable relative time string.
///
/// # Examples
///
/// ```
/// use chrono::{Utc, Duration};
/// use ithil::utils::format_relative_time;
///
/// let now = Utc::now();
/// assert_eq!(format_relative_time(now), "just now");
///
/// let five_min_ago = now - Duration::minutes(5);
/// assert_eq!(format_relative_time(five_min_ago), "5m ago");
/// ```
#[must_use]
pub fn format_relative_time(time: DateTime<Utc>) -> String {
    let now = Utc::now();
    let diff = now.signed_duration_since(time);

    // Future times
    if diff < Duration::zero() {
        let abs_diff = -diff;
        return format_future_time(abs_diff, time);
    }

    // Past times
    format_past_time(diff, time)
}

/// Format a time in the future.
fn format_future_time(diff: Duration, time: DateTime<Utc>) -> String {
    if diff < Duration::minutes(1) {
        "in a moment".to_string()
    } else if diff < Duration::hours(1) {
        format!("in {}m", diff.num_minutes())
    } else if diff < Duration::hours(24) {
        format!("in {}h", diff.num_hours())
    } else {
        format_absolute_time(time)
    }
}

/// Format a time in the past.
fn format_past_time(diff: Duration, time: DateTime<Utc>) -> String {
    if diff < Duration::minutes(1) {
        "just now".to_string()
    } else if diff < Duration::hours(1) {
        let minutes = diff.num_minutes();
        if minutes == 1 {
            "1m ago".to_string()
        } else {
            format!("{minutes}m ago")
        }
    } else if diff < Duration::hours(24) {
        let hours = diff.num_hours();
        if hours == 1 {
            "1h ago".to_string()
        } else {
            format!("{hours}h ago")
        }
    } else if diff < Duration::days(7) {
        let days = diff.num_days();
        if days == 1 {
            "yesterday".to_string()
        } else {
            format!("{days}d ago")
        }
    } else if diff < Duration::days(30) {
        let weeks = diff.num_weeks();
        if weeks == 1 {
            "1w ago".to_string()
        } else {
            format!("{weeks}w ago")
        }
    } else {
        format_absolute_time(time)
    }
}

/// Formats a time as an absolute string.
///
/// Returns different formats based on how recent the time is:
/// - Today: "15:04"
/// - Yesterday: "Yesterday 15:04"
/// - This year: "Jan 02 15:04"
/// - Other: "Jan 02, 2006 15:04"
#[must_use]
pub fn format_absolute_time(time: DateTime<Utc>) -> String {
    let local_time = time.with_timezone(&Local);
    let now = Local::now();

    if is_today(&local_time, &now) {
        return local_time.format("%H:%M").to_string();
    }

    if is_yesterday(&local_time, &now) {
        return format!("Yesterday {}", local_time.format("%H:%M"));
    }

    if local_time.format("%Y").to_string() == now.format("%Y").to_string() {
        return local_time.format("%b %d %H:%M").to_string();
    }

    local_time.format("%b %d, %Y %H:%M").to_string()
}

/// Formats a duration in a human-readable way.
///
/// # Arguments
///
/// * `duration` - The duration to format
///
/// # Returns
///
/// A formatted string like "5m", "2h 30m", etc.
///
/// # Examples
///
/// ```
/// use chrono::Duration;
/// use ithil::utils::format_duration;
///
/// assert_eq!(format_duration(Duration::seconds(30)), "30s");
/// assert_eq!(format_duration(Duration::minutes(5)), "5m");
/// assert_eq!(format_duration(Duration::hours(2)), "2h");
/// ```
#[must_use]
pub fn format_duration(duration: Duration) -> String {
    let total_seconds = duration.num_seconds();

    if total_seconds < 60 {
        return format!("{total_seconds}s");
    }

    if total_seconds < 3600 {
        let minutes = duration.num_minutes();
        let seconds = total_seconds % 60;
        if seconds == 0 {
            return format!("{minutes}m");
        }
        return format!("{minutes}m {seconds}s");
    }

    let hours = duration.num_hours();
    let minutes = (total_seconds / 60) % 60;
    if minutes == 0 {
        return format!("{hours}h");
    }
    format!("{hours}h {minutes}m")
}

/// Checks if a datetime is today.
fn is_today<Tz: chrono::TimeZone>(time: &DateTime<Tz>, now: &DateTime<Local>) -> bool {
    let time_local = time.with_timezone(&Local);
    time_local.date_naive() == now.date_naive()
}

/// Checks if a datetime is yesterday.
fn is_yesterday<Tz: chrono::TimeZone>(time: &DateTime<Tz>, now: &DateTime<Local>) -> bool {
    let time_local = time.with_timezone(&Local);
    let yesterday = now
        .date_naive()
        .pred_opt()
        .unwrap_or_else(|| now.date_naive());
    time_local.date_naive() == yesterday
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn just_now() {
        let now = Utc::now();
        assert_eq!(format_relative_time(now), "just now");
    }

    #[test]
    fn minutes_ago() {
        let time = Utc::now() - Duration::minutes(5);
        assert_eq!(format_relative_time(time), "5m ago");
    }

    #[test]
    fn one_minute_ago() {
        let time = Utc::now() - Duration::minutes(1);
        assert_eq!(format_relative_time(time), "1m ago");
    }

    #[test]
    fn hours_ago() {
        let time = Utc::now() - Duration::hours(3);
        assert_eq!(format_relative_time(time), "3h ago");
    }

    #[test]
    fn one_hour_ago() {
        let time = Utc::now() - Duration::hours(1);
        assert_eq!(format_relative_time(time), "1h ago");
    }

    #[test]
    fn yesterday() {
        let time = Utc::now() - Duration::days(1);
        assert_eq!(format_relative_time(time), "yesterday");
    }

    #[test]
    fn days_ago() {
        let time = Utc::now() - Duration::days(3);
        assert_eq!(format_relative_time(time), "3d ago");
    }

    #[test]
    fn weeks_ago() {
        let time = Utc::now() - Duration::weeks(2);
        assert_eq!(format_relative_time(time), "2w ago");
    }

    #[test]
    fn future_time() {
        // Use 10 minutes and check for a range to avoid race conditions
        // at minute boundaries
        let time = Utc::now() + Duration::minutes(10);
        let result = format_relative_time(time);
        // Should be "in 9m" or "in 10m" depending on exact timing
        assert!(
            result == "in 9m" || result == "in 10m",
            "Expected 'in 9m' or 'in 10m', got '{result}'"
        );
    }

    #[test]
    fn format_duration_seconds() {
        assert_eq!(format_duration(Duration::seconds(30)), "30s");
        assert_eq!(format_duration(Duration::seconds(59)), "59s");
    }

    #[test]
    fn format_duration_minutes() {
        assert_eq!(format_duration(Duration::minutes(5)), "5m");
        assert_eq!(format_duration(Duration::seconds(90)), "1m 30s");
    }

    #[test]
    fn format_duration_hours() {
        assert_eq!(format_duration(Duration::hours(2)), "2h");
        assert_eq!(format_duration(Duration::minutes(150)), "2h 30m");
    }
}
